rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1152,6 +1156,64 @@ rmw_context_impl_t::fini()
   return RMW_RET_OK;
 }
 
+#ifdef DDS_HAS_SHM
+template<typename entityT>
+static void * init_and_alloc_sample(
+  entityT & entity, const uint32_t sample_size, const bool alloc_on_heap = false)
+{
+  // initialise the data allocator
+  if (alloc_on_heap) {
+    RET_EXPECTED(
+      dds_data_allocator_init_heap(&entity->data_allocator),
+      DDS_RETCODE_OK,
+      "Reader data allocator initialization failed for heap",
+      return nullptr);
+  } else {
+    RET_EXPECTED(
+      dds_data_allocator_init(entity->enth, &entity->data_allocator),
+      DDS_RETCODE_OK,
+      "Writer allocator initialisation failed",
+      return nullptr);
+  }
+  // allocate memory for message + header
+  auto chunk_ptr = dds_data_allocator_alloc(
+    &entity->data_allocator,
+    DETERMINE_ICEORYX_CHUNK_SIZE(sample_size));
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    chunk_ptr,
+    "Failed to get loan",
+    return nullptr);
+  auto ice_hdr = static_cast<iceoryx_header_t *>(chunk_ptr);
```

Comment:
Cyclone really needs a bit of fixing up so that there is no need to rely on definitions that are conditional on `DDS_HAS_SHM`. At least I think the only reason there is still `#ifdef DDS_HAS_SHM` in here is:
* `DETERMINE_ICEORYX_CHUNK_SIZE`
* `SHIFT_PAST_ICEORYX_HEADER`
* `SHIFT_BACK_TO_ICEORYX_HEADER`
* `iceoryx_header_t`
* `iox_chunk` in `struct ddsi_serdata`
* `iox_size` in `struct ddsi_sertype`

There's general agreement that the Cyclone integration of Iceoryx needs refactoring and generalised a little bit. In my view, that would be the right time to clean this up as well.

Reply:
I agree. Also to add to the above list just to be complete, there is also the conditional definitions for iox functions in `ddsi_serdata_ops` (which can ideally also be generalized)